src/actions/write/initiateLeave.ts

```diff
@@ -0,0 +1,39 @@
+import { NETWORK_LEAVING_STATUS } from '../../constants';
+import { ContractResult, IOState, PstAction } from '../../types';
+
+declare const ContractError;
+declare const SmartWeave: any;
+
+// Begins the network leave process for a gateway operator
+export const initiateLeave = async (
+  state: IOState,
+  { caller }: PstAction,
+): Promise<ContractResult> => {
+  const settings = state.settings;
+  const gateways = state.gateways;
+
+  if (!(caller in gateways)) {
+    throw new ContractError('This target is not a registered gateway.');
+  }
+
+  if (gateways[caller].status === NETWORK_LEAVING_STATUS) {
+    throw new ContractError(
+      'This Gateway is in the process of leaving the network',
+    );
+  }
+
+  if (
+    gateways[caller].start + settings.minGatewayJoinLength >
+    +SmartWeave.block.height
+  ) {
+    throw new ContractError('This Gateway has not been joined long enough');
+  }
+
```

Comment:
@vilenarios - i believe we want to update the `end` for all vaults within the respective delegate.

Reply:
added in [804323e](https://github.com/ar-io/arns-pilot/pull/21/commits/804323e2d247207fee45947075ca60d46e0269e5)